staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -119,8 +119,10 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 
 	fs.StringSliceVar(&s.WatchCacheSizes, "watch-cache-sizes", s.WatchCacheSizes, ""+
 		"List of watch cache sizes for every resource (pods, nodes, etc.), comma separated. "+
-		"The individual override format: resource#size, where size is a number. It takes effect "+
-		"when watch-cache is enabled.")
+		"The individual override format: resource[.group]#size, where resource is lowercase plural (no version), "+
+		"group is optional, and size is a number. It takes effect when watch-cache is enabled. "+
```

Comment:
Based on the code in https://github.com/kubernetes/apiserver/blob/f5fd0005b06697e897cbeb10c97b7cb5f5b85232/pkg/server/options/etcd.go#L194-L203 it seems to me that although the parser considers the group to be optional, the larger context requires that the group be specified.

And I wonder what is the right thing to do for resources in the default group.

Reply:
group is optional, when group is not present, the resource is treated as in legacy/v1 group

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/api-group.md#bookkeeping-for-groups

"For backward compatibility, v1 objects belong to the group with an empty name, so existing v1 config files will remain valid."

Reply:
while the group is optional as far as the parser is concerned, and perhaps other code as well, a brief glance at the code in https://github.com/kubernetes/apiserver/blob/f5fd0005b06697e897cbeb10c97b7cb5f5b85232/pkg/server/options/etcd.go#L194-L203 shows that the group is not optional.  `cacheSize, ok := sizes[resource]` is used to look up a particular resource.  Only one key is sought in the `sizes` map.  Other keys may be present, but they will have no effect.  BTW, that leads to a bad UX.  A mistaken key leads to a silent misconfiguration.

Reply:
the code you mentioned (https://github.com/kubernetes/apiserver/blob/f5fd0005b06697e897cbeb10c97b7cb5f5b85232/pkg/server/options/etcd.go#L194-L203) is called by (https://github.com/kubernetes/apiserver/blob/73f85001b53a0adc72b5d3dc68abbcda0837fe3a/pkg/registry/generic/registry/store.go#L1288) where e.DefaultQualifiedResource is passed.

For resources in core group (v1/legacy), the group name is "" (https://github.com/kubernetes/kubernetes/blob/a1921f665be1b8c6151f8131a4474ad660268430/pkg/apis/core/register.go#L26)

For example, for "pods", its DefaultQualifiedResource is defined as api.Resource("pods") (https://github.com/kubernetes/kubernetes/blob/dd9981d038012c120525c9e6df98b3beb3ef19e1/pkg/registry/core/pod/storage/storage.go#L71)

Reply:
So the cluster operator must use exactly the string present in `e.DefaultQualifiedResource`; any other string appearing in a watch cache sizes spec will be accepted and ignored.

I think you are telling me that for resources in the core group, `e.DefaultQualifiedResource` has the empty string as the group name.  I presume that for all the other groups, `e.DefaultQualifiedResource` does *not* have the empty string as the group name.

If I understand correctly, the operator needs to be told that --- in a watch cache sizes spec --- the group MUST be _omitted_ for resources in the default group and MUST NOT be omitted for resources in other groups.